Add editorUI instead of plain JEditorPane to have linenumbers

Chris2011 opened the pull request:
I added the editorUI instead of plain JEditorPane. We have some more options like showing line numbers and formatting.
Unfortunately we still have some other things to fix:

* Some sidebar actions are available, which are not needed (Show linenumbers, show bookmarks)
* Some editor actions for specific mimetype are there (navigate, insert code, etc.)
* We have the breadCrumb bar below the editor, which is also not needed
* Sidebar (Linenumbers) and Breadcrumb bar are synced with the normal editor. If we deactivate the linenumbers for any editor, or inside our editor it is synced with all other editors same for showing breadcrumbs

I dunno whether we can hack those things for example add a new popupMenu with couple of actions (copy, format, etc.). I tried this but it didn't work out.

So at the end it is up to use, to merge this or we should just add simple line numbers to JEditorPane as the editor already does.

Also I fixed a JSON parsing exception.

javierllorente commented:
Perhaps we could completely disable sidebar and editor actions for our EditorUI. There are no actions in the current version.
Same for the breadcrumb.

Chris2011 commented:
I dunno what sidebar means here. The thing between linenumbers and the code? I tried to disable/remove those things but I failed as mentioned. Maybe you have an idea here.

Chris2011 commented:
If it is the linenumbers, remove them is not what the PR is. The PR added the LineNumbers, this is why I exactly did this.

javierllorente commented:
I mean where the line numbers are. If you right click there, you get a menu (Bookmark, Breakpoint, ...).
EditorUI has nice details; line numbers, line connecting the start and end of curly braces, going back to the current line, ...
There is probably a method to disable the mentioned right click menus. I have to take a look at it.

Chris2011 commented:
Ok, now it is clear. Thx :)

Chris2011 commented:
As I tried a bit more, if you open the diff panel, they use JEditorPane which also has lineNumbers and there you don't have any context menu on the line numbers. I didn't dive into the code yet but I guess we have 2 options:

* Prevent the contextmenu somehow (Have a look into the diff core code how they do it)
* Make a sidebar with linenumbers by our own (Also have a look into the diff core code, maybe they already did this)

Chris2011 commented:
Ok, so it seems, that they are creating the linenumbers and not using the editorkit: https://github.com/apache/netbeans/blob/master/ide/diff/src/org/netbeans/modules/diff/builtin/visualizer/LinesComponent.java#L138 I will rewrite my code then.

javierllorente commented:
I haven't looked deeply into it. Thanks for sharing your findings!
Perhaps we could replicate somehow createLines()

Chris2011 commented:
This is what I'm trying, but it is much more code, than just the createLines function. I'm on it and I'm trying it.

Chris2011 commented:
We need much more methos like paintComponent, other methods for the lineHeight and linenumbers, etc. So it is much more work unfortunately but it is a good start. Also if I can encapsulate this I can reuse it in other plugins/ways ;)

javierllorente commented:
It would be nice if we could just override some method to disable the right click menu and get the features we want (line numbers, etc).
Not sure if reimplementing createLines() and friends is the right path.

Chris2011 commented:
I really don't think that this is possible, this is why they reimplemented the LineNumbers inside of the diff window. Just guessing, but if this would be possible there would be an option. Unfortunately a JEditorPane just can handle mimeType but doesn't have a flag like showLineNumbers(). This is used via the editorKit which makes much more, which we don't need or we need to recreate it.

Chris2011 commented:
And if we can recreates it we can create a JEditorPane2 or JEditorPaneAdvanced or wahtever.

javierllorente commented:
So, this is a case of either take it all or nothing. 
Let's brake it down:
- Line numbers
- Line connecting start/end tag or curly braces for XML/JSON
Anything else?

Chris2011 commented:
We need a sidebar for linenumbers. Each line is one number like in the current editor and in the diff tool. Add new linenumber when adding new line, remove line number, when removing last line. So same behavior, we just need to do it and I'm on it.

> Line connecting start/end tag or curly braces for XML/JSON
This I would se it in an other task. We should do it step by step but also cool feature, yes.

javierllorente commented:
Sure. Let's do it step by step. 
Thanks!

Chris2011 commented:
Hey @javierllorente finally I made it, a simple one from scratch, not using the one LineComponents from the DiffPanel, it was to much.

We still have two problems, but we can handle this later. I added comments to the LineNumberComponent:

1. When you zoom the response editor it will not change the size of the linenumber yet
2. The linenumber component has a fixed size. I added a method called adjustWidth and I tried to use it and it worked but the problem was that the editor scroll was not aligned anymore with the linenumber component, it was a bit async.

So those are 2 little problems but needs to be handled and we should handle it later to how have linenumbers. Better than nothing.

javierllorente commented:
Looks good!
In my case, there is a minor detail: the line number font isn't anti-aliased, so it doesn't look smooth. Out of curiosity, does it happen to you too?
I think that those problems you mention are little.

Chris2011 commented:
Will have a look thx. In my test it fits to the editor font. So let's see what I can do.

javierllorente commented:
Thanks to you!
Perhaps it's me (running KDE on Linux).
Here's the screenshot:

![imagen](https://github.com/javierllorente/netbeans-rest-client/assets/1075241/c001df4b-5ab0-4fca-91a6-6e272e558ba4)

javierllorente commented:
BTW, LineNumberComponent is missing the license header. Please add it with your name.

Chris2011 commented:
@javierllorente done, please test it :)

javierllorente commented:
Tested. Thanks a lot! :)
